lib/streamlit/caching.py

```diff
@@ -527,6 +479,13 @@ def wrapped_func(*args, **kwargs):
             message = "Running `%s(...)`." % name
 
         def get_or_create_cached_value():
+            nonlocal cache_key
+            if cache_key is None:
+                # Wait with generating the cache key until the first call.
+                # This way we can see complete globals, including functions
+                # defined after this one.
```

Comment:
Very minor thing, but it would be nice to mention here that if the hash is computed earlier, functions (and other globals) defined after the cached function will only have their names included in the hash.

Reply:
Would this be better?

                # Delay generating the cache key until the first call.
                # This way we can see values of globals, including functions
                # defined after this one.
                # If we generated the key earlier we would only hash those
                # globals by name, and miss changes in their code or value.

Reply:
Yep, that's perfect, thanks!